sklearn/preprocessing/data.py

```diff
@@ -2369,16 +2389,47 @@ def _transform_col(self, X_col, quantiles, inverse):
 
         return X_col
 
+    def _assert_finite_or_nan(self, X):
+        """Check that X contain finite or NaN values."""
+        X = np.asanyarray(X)
+        # First try an O(n) time, O(1) space solution for the common case that
+        # everything is finite; fall back to O(n) space np.isfinite to prevent
+        # false positives from overflow in sum method.
+        if (X.dtype.char in np.typecodes['AllFloat']
+                and not np.isfinite(X[~np.isnan(X)].sum())
+                and not np.isfinite(X[~np.isnan(X)]).all()):
+            raise ValueError("Input contains infinity"
+                             " or a value too large for %r." % X.dtype)
+        if np.count_nonzero(self._mask_missing_values):
+            if LooseVersion(np.__version__) >= '1.9':
+                self._percentile_func = np.nanpercentile
+            else:
+                raise NotImplementedError(
+                    'QuantileTransformer does not handle NaN value with'
+                    ' NumPy {}. Please upgrade to NumPy 1.9. or higher'.format(
+                        np.__version__))
+        else:
+            self._percentile_func = np.percentile
+
     def _check_inputs(self, X, accept_sparse_negative=False):
         """Check inputs before fit and transform"""
+        if sparse.issparse(X):
```

Comment:
Why before check_array?

Reply:
`check_array` will convert the matrix into a float dtype. I wanted to compare when data can still be int.
However I could also use `np.isclose` which can handle NaN as well.